src/Data/Text/Builder/Linear/Dec/Unbounded.hs

```diff
@@ -0,0 +1,380 @@
+unsafePrependDec marr off@(I# off#) n = case n of
+  I.IS i# → Bounded.unsafePrependDec marr off (I# i#)
+  _ → unsafePrependBigNatDec marr (off# -# 1#) (integerToBigNat# n) >>= prependSign
+    where
+      prependSign !off' =
+        if n < 0
+          then do
+            A.unsafeWrite marr (off' - 1) 0x2d -- '-'
+            pure (off - off' + 1)
+          else pure (off - off')
+{-# INLINEABLE unsafePrependDec #-}
+
+type DigitsWriter s = Int# → BN.BigNat# → ST s Int
+
+-- TODO: Remove once we choose a strategy
+data Strategy = SmallOnly | BigOnly | HugeOnly | Optimum
+
+-- Use the fastest writer depending on the BigNat size
+unsafePrependBigNatDec ∷ ∀ s. A.MArray s → DigitsWriter s
+unsafePrependBigNatDec marr !off0 !n0
+  | BN.bigNatSize n0 < bigSizeThreshold = prependSmallNat marr off0 n0
+  | BN.bigNatSize n0 < hugeSizeThreshold = prependBigNat marr off0 n0
+  | otherwise = prependHugeNat marr off0 n0
+  where
+    -- TODO: simplify once we choose a strategy
+    -- Strategy on the 'BigNat#' size in order to scale
+    -- strategy = SmallOnly
+    -- strategy = BigOnly
+    -- strategy = HugeOnly
+    strategy = Optimum
+    bigSizeThreshold, hugeSizeThreshold ∷ Word
+    !(bigSizeThreshold, hugeSizeThreshold) = case strategy of
+      SmallOnly → (maxBound, maxBound)
+      BigOnly → (minBound, maxBound)
+      HugeOnly → (minBound, minBound)
+      Optimum → (25, 400)
+
+-- Writer for “small” 'BigNat's.
+--
+-- Divide repeatedly by poweredBase.
+prependSmallNat ∷ ∀ s. A.MArray s → DigitsWriter s
+prependSmallNat marr = go
+  where
+    !(# power, poweredBase, _poweredBase² #) = selectPower (# #)
+
+    go ∷ DigitsWriter s
+    go !o1 !n = case n `BN.bigNatQuotRemWord#` poweredBase of
+      (# q, r #) → do
+        !o2 ← unsafePrependWordDec marr (I# o1) (W# r)
+        if BN.bigNatIsZero q
+          then pure o2
+          else do
+            let !o3 = o1 -# (word2Int# power -# 1#)
+            padWithZeros marr (I# o3) (o2 - I# o3)
+            go (o3 -# 1#) q
+
+-- Writer for “big” 'BigNat's.
+--
+-- Divide repeatedly by pow10^k, where pow10^k ≈ √n, then by pow10.
+prependBigNat ∷ ∀ s. A.MArray s → DigitsWriter s
+prependBigNat marr off0 n0 = prependBig off0 n0
+  where
+    !(# power, poweredBase, poweredBase² #) = selectPower (# #)
+    !(# bigBaseCount, bigBase #) = mkBigBase n0
+
+    prependBig ∷ DigitsWriter s
+    prependBig !o1 !n = case n `BN.bigNatQuotRem#` bigBase of
+      (# q, r #) →
+        if BN.bigNatIsZero q
+          then prependSmallNat marr o1 r
+          else do
+            I# o2 ← prependSmall bigBaseCount o1 r
+            prependBig (o2 -# 1#) q
+
+    -- Find k such that pow10^k ≈ √n
+    -- We start at (2, poweredBase²) and not (1, poweredBase), because the latter case
+    -- is handled by `prependSmallNat`.
+    mkBigBase ∷ BN.BigNat# → (# Word#, BN.BigNat# #)
+    mkBigBase n# = go 2## poweredBase²
+      where
+        !targetLen = double2Int# (sqrtDouble# (int2Double# (BN.bigNatSize# n#)))
```

Comment:
I don't quite follow this. Why we take square root instead of taking half?

Reply:
Borrowed from [Rust `bigint`](https://github.com/rust-num/num-bigint/blob/124eac8f0385197c12e22ed8a767ed268e71f0b0/src/biguint/convert.rs#L708).

Reply:
In such case `pow10^k` does not seem to be anywhere near `sqrt n`, as the comment says. Or am I deluded?..

Reply:
I fixed the note to:
> Find k such that bigNatSize (pow10^k) ≈ √(bigNatSize n)